File: Remove Duplicates from Sorted Array.java

// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
/*Leveraging a 2 pointer approach.
The first pointer traverses throught the complete array and we keep track of the particular element where we are at.
The second pointer is used to keep trak of the location where the next valid should be
As soon as the element changes we set count to 1.
Also along with this keep copying element from the location of the first pointer to the location of the second pointer till the count is <=2.
*/
// Your code here along with comments explaining your approach
class Solution {
    public int removeDuplicates(int[] nums) {
        if(nums == null || nums.length == 0){
            return 0;
        }
        int j = 1;
        int count = 1;
        for(int i = 1; i < nums.length ; i++){
            if(nums[i] == nums[i-1]){
                count++;
            }else{
                count = 1;
            }
            if(count <=2){
                nums[j] = nums[i];
                j++;
            }
        }
        return j;
    }
}

File: removeDuplicatesFromSortedArray.py
# // Time Complexity : O(n)
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode : Yes
# // Three line explanation of solution in plain english

class solution:
    def removeDuplicates(self, nums):
        ## this is a tricky question. here what we do is have a slow pointer starting from 1 and the count also as 1 .
        ## this is because the first element is always unique , so we start from the next. now we iteratore from index 1 
        ## to end and see if nums[i] == nums[i-1] this means that we are checking so that we know if we can increase count or 
        ## reinitialize it to 1. Now we check if this count <= 2(k) if it is so we get the value at nums[i] and place it in nums[slow]
        ## and increment the slow pointer. At the end we will return slow. which is the end the array.

        slow = 1
        count = 1
        k = 2
        for i in range(1,len(nums)):
            if nums[i] != nums[i-1]:
                count = 1
            else:
                count+=1
            
            if count <= k:
                nums[slow] = nums[i]
                slow+=1

        return slow

        
